### PR TITLE
Upgrade wiki.spawn_groups onto models and clean the output up

### PR DESCRIPTION
**Pull request by alex4401**
tl;dr These changes get rid of some bad design choices back from when first iteration of the maps exported was still in development (mainly mimicking Q-data to help existing scripts).

- Format version is bumped from 3 to 4.
- Breaks what's left of compatibility with Qowyn's data.
- Limits for nulls are no longer exported.
- Consistency update: change `blueprintPath` to `bp`.
- Condensed limit model a little: `desiredNumberMult` changed to a simple `mult` with an appropriate field description in the schema.
- Zipped up `classes`, `classWeights` and `spawnOffsets` in an NPC group. Each entry is now an object. Offset field defaults to a (0, 0, 0) vector.
- Several Extinction group containers are now exported.
- Joined `weights` and `to` for swap rules into a single field `to`  of format `(weight, class)`.
- Top level fields renamed to be cleaner/intuitive and more technically correct:
- * spawngroups -> containers
- * classSwaps -> randomSwaps (not to be confused with `remaps`)
- * externalGroupChanges -> extraGroups (as it is not possible for a mod to remove a group)
- * dinoRemaps -> remaps (includes spawning group container remaps now)
- Random replacements of `null` no longer exported.
- Random group-level replacements where the only rule affects a non-existent class no longer exported.
- Limit specifiers that do not affect the result at all (i.e. multiplier=1) no longer exported.
- Cleaned up type on `WeighedClassSwap->to` (was Optional[Optional[str]]).
- _Due to a pydantic bug the resulting JSON schema incorrectly reports some nullable str fields as non-nullable._
- maps: Map-level replacements of null classes no longer exported.
- svgs: Better topomap offsets for Crystal Isles and Valguero
- maps: Export Caballus' Ice Wyvern nests.
- maps: Fixed a crash when an NPC manager doesn't meet requirements to be valid - hunted down in Caballus.
- maps: isHyper field on charge nodes.

### CODE REVIEW

File: export/wiki/spawn_groups/remaps.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+from automate.hierarchy_exporter import ExportModel, Field
+from ue.asset import UAsset
+from ue.utils import sanitise_output
+
+__all__ = ['convert_npc_remaps']
+
+
+class ClassRemap(ExportModel):
+    from_bp: str = Field(alias="from")
+    to: Optional[str] = None
```
Comment (coldino) [inline export/wiki/spawn_groups/remaps.py:12]: This looks reusable. Please move somewhere more central. Maybe create `export.models` or similar?

File: export/wiki/spawn_groups/structs.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        if v:
+            out.append(v)
+
+    return out
+
+
+class Vector(ExportModel):
+    x: float
+    y: float
+    z: float
```
Comment (coldino) [inline export/wiki/spawn_groups/structs.py:64]: This looks reusable. Please move somewhere more central. Maybe create `export.models` or similar?

File: export/wiki/spawn_groups/structs.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        species=list(),
+        randomSwaps=[],
+    )
+
+    # Export zipped NPC entries
+    chances = d['NPCsToSpawnPercentageChance'].values
+    offsets = d['NPCsSpawnOffsets'].values
+    for index, kls in enumerate(d['NPCsToSpawn'].values):
+        npc = NpcEntry(
+            chance=chances[index] if index < len(chances) else 1,
```
Comment (coldino) [inline export/wiki/spawn_groups/structs.py:99]: If this runs many many times then please cache `len(...)`.

File: export/wiki/spawn_groups/structs.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            bp=sanitise_output(kls),
+            offset=sanitise_output(offsets[index] if index < len(offsets) else Vector(x=0, y=0, z=0)),
```
Comment (coldino) [inline export/wiki/spawn_groups/structs.py:101]: If this runs many many times then please cache `len(...)`.

File: export/wiki/spawn_groups/structs.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+]
+
+
+def _zip_swap_outputs(d: Dict[str, Any]) -> Iterable[Tuple[float, Optional[str]]]:
+    npcs: ArrayProperty = d['ToClasses']
+    weights = d['Weights'].values
+
+    for index, kls in enumerate(npcs.values):
+        # Get weight of this class. Defaults to 1 if array is too short.
+        weight = float(weights[index]) if index < len(weights) else 1.0
```
Comment (coldino) [inline export/wiki/spawn_groups/structs.py:26]: If this runs many many times then please cache `len(...)`.

File: export/wiki/spawn_groups/structs.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        yield (weight, sanitise_output(kls))
```
Comment (coldino) [inline export/wiki/spawn_groups/structs.py:27]: Did you look at Python's `zip_longest` as an alternative here? Your approach may be neater given the extra work needed, but wanted you to be aware that the function exists.

**Review by coldino — COMMENTED** [review]
Looking pretty good! Just minor queries before merge, really.

File: export/wiki/spawn_groups/structs.py (hunk 3 above)
Comment (alex4401) [reply export/wiki/spawn_groups/structs.py:99]: It shouldn't run an insane number of times (though it is dependent on the input - I do not believe we have a mod on our list that would spawn two hundred dinos in one spawn pass though), but I will cache it, thank you.
